src/bases/D3MPoolBase.sol

```diff
@@ -0,0 +1,124 @@
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity 0.6.12;
+
+interface TokenLike {
+    function approve(address, uint256) external returns (bool);
+    function transfer(address, uint256) external returns (bool);
+    function transferFrom(address, address, uint256) external returns (bool);
+    function balanceOf(address) external view returns (uint256);
+}
+
+interface CanLike {
+    function hope(address) external;
+    function nope(address) external;
+}
+
+interface d3mHubLike {
+    function vat() external view returns (address);
+}
+
+interface D3MPlanLike {
+    function calcSupplies(uint256) external view returns (uint256, uint256);
+    function maxBar() external view returns (uint256);
+}
+
+abstract contract D3MPoolBase {
+    // --- Auth ---
+    mapping (address => uint256) public wards;
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+    modifier auth {
+        require(wards[msg.sender] == 1, "D3MPoolBase/not-authorized");
+        _;
+    }
+
+    TokenLike   public immutable asset; // Dai
+    address     public immutable hub;
+    address     public immutable pool;
+
+    address     public           share;
+    address     public           plan; // How we calculate target debt
+    uint256     public           live = 1;
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event Cage();
+    // --- EIP-4626 Events ---
+    event Deposit(address indexed caller, address indexed owner, uint256 assets, uint256 shares);
+    event Withdraw(address indexed caller, address indexed receiver, address indexed owner, uint256 assets, uint256 shares);
+
+    constructor(address hub_, address dai_, address pool_) internal {
+
+        pool = pool_;
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+
+        asset = TokenLike(dai_);
+
+        hub = hub_;
+        wards[hub_] = 1;
+        emit Rely(hub_);
+        CanLike(d3mHubLike(hub_).vat()).hope(hub_);
+    }
+
+    // --- Admin ---
+    function file(bytes32 what, address data) public virtual auth {
+        require(live == 1, "D3MPoolBase/no-file-not-live");
+
+        if (what == "share") {
+            if (share != address(0)) TokenLike(share).approve(hub, 0);
+            share = data;
+            TokenLike(data).approve(hub, type(uint256).max);
+        } else if (what == "plan") plan = data;
+    }
+
+    function validTarget() external view virtual returns (bool);
+
+    function calcSupplies(uint256 availableLiquidity) external view virtual returns (uint256, uint256);
+
+    function deposit(uint256 amt) external virtual;
+
+    function withdraw(uint256 amt) external virtual;
+
+    function collect(address[] memory assets, uint256 amount) external virtual returns (uint256 amt); // should use auth
```

Comment:
Here the return parameter is explicitly named as opposed to the other functions.

Reply:
removed function from base completely so pools can independently define the interface needed